File: 42-trapping-rain-water/42-trapping-rain-water.ts
/*
[0,1,0,2,1,0,1,3,2,1,2,1]
[4,2,0,3,2,5]
[4,2,3]
[6,4,2,0,3,2,0,3,1,4,5,3,2,7,5,3,0,1,2,1,3,4,6,8,1,3]
*/
function trap(H:number[]): number {

  const N = H.length;
  let l = 0,
      r = N - 1,
      lMax = 0,
      rMax = N - 1,
      ans = 0;
  
  while(l <= r) {
    if(H[l] <= H[r]) {
      if(H[l] >= H[lMax]) {
        lMax = l;
      } else {
        ans += H[lMax] - H[l];
      }
      l++;
    } else {
      if(H[r] >= H[rMax]) {
        rMax = r;
      } else {
        ans += H[rMax] - H[r];
      }
      r--;
    }   
  }
  return ans;
}

function trapMy(height: number[]): number {
  const N = height.length,
        mono = new Array<number>(N),
        dp = new Array<number>(N).fill(-1);
  
  let ans = 0;
  
  for(let i = 0; i < N; i++) {
    let max = i;
    
    for(let j = i + 1; j < N; j++) {
      if(height[max] == height[j])
        max = j;
      
      if(height[j] > height[max]){
        max = j;
        break;
      }
    }
    
    mono[i] = max;
  }
  
  mono.forEach((m, i) => console.log(i, height[i], m, height[m]))

  if(height[0] != 0)
    dfs(0, mono[0]);
  else
    dfs(1, mono[1]);

  
  function dfs(l: number, r: number) {
    if(r - l <= 1 || height[l] == 0 || l >= N || r >= N) {
      return;
    }

    const minH = Math.min(height[l], height[r]),
          diff = r - l - 1,
          totalWater = minH * diff,
          buildings = calcBuildingHeightInBetween(l + 1, r);

    ans += totalWater - buildings;

    console.log({l, r, minH, diff, totalWater, buildings, ans})
    
    for(let i = r; i < N; i++) {
      if(i == mono[i])
        continue;
      else
        return dfs(i, mono[i]);
    }
  }
  
  function calcBuildingHeightInBetween(l: number, r: number): number {
    let ans = 0;
    for(let i = l; i < r; i++)
      ans += height[i];
    return ans;
  }
  
  return ans;
};

File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        return trap_2Pointer(height);
    }    

    public int trap_2Pointer(int[] height) {
        int N = height.length, ans = 0;
        int l = 0, r = N - 1;
        int lMax = 0, rMax = 0;

        while (l < r)
            if (height[l] < height[r])
                if (height[l] > lMax)
                    lMax = height[l++];
                else
                    ans += lMax - height[l++];
            else
                if (height[r] > rMax)
                    rMax = height[r--];
                else
                    ans += rMax - height[r--];

        return ans;
    }
    
    public int trap_Stack(int[] height) {
        int N = height.length, ans = 0;
        Stack<Integer> mono = new Stack<>();

        for (int i = 0; i < N; i++) {
            while (!mono.isEmpty() && height[i] > height[mono.peek()]) {
                int idx = mono.pop();
                if (mono.isEmpty())
                    break;
                int dist = i - mono.peek() - 1;
                int boundedHeight = Math.min(height[i], height[mono.peek()]) - height[idx];

                ans += dist * boundedHeight;
            }
            mono.push(i);
        }

        return ans;
    }
    
    public int trap_DP(int[] height) {
        int N = height.length, ans = 0;

        if (N == 0) return ans;

        int[] leftMax = new int[N],
                rightMax = new int[N];

        leftMax[0] = height[0];
        rightMax[N - 1] = height[N - 1];

        for (int i = 1; i < N; i++)
            leftMax[i] = Math.max(leftMax[i - 1], height[i]);

        for (int i = N - 2; i >= 0; i--)
            rightMax[i] = Math.max(rightMax[i + 1], height[i]);

        for (int i = 1; i < N - 1; i++)
            ans += Math.min(leftMax[i], rightMax[i]) - height[i];

        return ans;
    }

    public int trap_again(int[] height) {
        int N = height.length;
        int ans = 0;
        Stack<Integer> stack = new Stack<>();
        
        for(int i = 0; i < N; i++) {
            while(!stack.isEmpty() && height[i] > height[stack.peek()]) {
                int idxOfLastMaxHeight = stack.pop();
                
                if(stack.isEmpty())
                    break;
                
                int distance = (i - 1) - stack.peek();
                int boundedHeight = Math.min(height[i], height[stack.peek()]) - height[idxOfLastMaxHeight];
                ans += distance * boundedHeight;
            }
            
            stack.push(i);           
        }
        
        return ans;
    }
}